File: packages/host4_flutter_aivoice/ios/AiVoice_副本/Tools/SubtitleMsgData+Validation.m
//
//  SubtitleMsgData+Validation.m
//  GameMacro
//
//  判断 subv 消息是否为完整句子的扩展
//

#import "SubtitleMsgData+Validation.h"

@implementation SubtitleMsgData (Validation)

#pragma mark - 机器人消息判断

- (BOOL)isBotMessageWithBotIds:(NSArray<NSString *> *)botUserIds {
    if (!self.userId || botUserIds.count == 0) {
        return NO;
    }
    
    return [botUserIds containsObject:self.userId];
}

- (BOOL)isBotMessage:(NSString *)botUserId {
    if (!self.userId || !botUserId) {
        return NO;
    }
    
    return [self.userId isEqualToString:botUserId];
}

#pragma mark - 完整句子判断

- (BOOL)isCompleteSentence {
    // definite = true 表示完整的分句
    // paragraph = true 表示段落结束（也是完整句子）
//    if ([self.userId isEqualToString:AIVoiceShared.userId]){
//        //用户回复
//        return self.definite;
//    }else{
//        //机器人回复
//        return self.paragraph;
//    }
    return self.definite;
}

- (BOOL)isParagraphEnd {
    return self.paragraph;
}

#pragma mark - 特殊内容判断

- (BOOL)isEmptyOrNil {
    if (!self.text) {
        return YES;
    }
    
    // 去除空白字符后检查
    NSString *trimmed = [self.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    return trimmed.length == 0;
}

- (BOOL)isSpecialSymbol {
//    if (!self.text) {
//        return NO;
//    }
    
    return [self.text isEqualToString:@"~"];
    
    NSString *trimmed = [self.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    
    // 判断是否为特殊符号
    NSArray<NSString *> *specialSymbols = @[@"~"];
    
    for (NSString *symbol in specialSymbols) {
        if ([trimmed isEqualToString:symbol]) {
            return YES;
        }
    }
    
    return NO;
}

- (BOOL)needsSpecialHandling {
    return [self isSpecialSymbol];
}

#pragma mark - 综合判断

- (BOOL)isBotCompleteSentenceNeedsSpecialHandling:(NSString *)botUserId {
    // 1. 必须是机器人消息（不是用扣费）
    if (![self isBotMessage:botUserId]) {
        return NO;
    }
    
    //特殊字符不扣费
    NSString *trimmed = [self.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    
    //2.判断是否为特殊符号
    NSArray<NSString *> *specialSymbols = @[@"~"];
    
    for (NSString *symbol in specialSymbols) {
        if ([trimmed isEqualToString:symbol]) {
            //是特殊字符，不需要扣费
            AIVoiceShared.isUserAsk = NO;
            
            NSLog(@"当前isUserAsk状态%d",AIVoiceShared.isUserAsk);
            
            return NO;
        }
    }
    
    //3. 不是完整内容，不扣费
    if (!self.paragraph) {
        return NO;
    }
    
//    //空字段不扣费
//    if (isStringEmpty(self.text)){
//        return NO;
//    }
    
    return YES;
}

- (BOOL)isBotCompleteSentenceAndPlayAudio:(NSString *)botUserId{
    // 1. 必须是机器人消息
    if (![self isBotMessage:botUserId]) {
        return NO;
    }
    
    // 2. 必须是完整句子
    if (![self isCompleteSentence]) {
        return NO;
    }
    
    // 3. 内容需要特殊处理（为空或特殊符号）
    return [self needsSpecialHandling];
}

- (SubtitleSentenceType)sentenceType {
    // 不完整句子
    if (![self isCompleteSentence]) {
        return SubtitleSentenceTypeIncomplete;
    }
    
    // 段落结束
    if ([self isParagraphEnd]) {
        return SubtitleSentenceTypeParagraphEnd;
    }
    
    // 内容为空
    if ([self isEmptyOrNil]) {
        return SubtitleSentenceTypeEmpty;
    }
    
    // 只有特殊符号
    if ([self isSpecialSymbol]) {
        return SubtitleSentenceTypeSpecialSymbol;
    }
    
    // 正常完整句子
    return SubtitleSentenceTypeComplete;
}

@end

File: packages/host4_flutter_aivoice/ios/Classes/Models/SubtitleMsgData+Validation.swift
import Foundation

// MARK: - 句子类型枚举
enum SubtitleSentenceType: Int {
  case incomplete   = 0  /// 不完整句子
  case complete     = 1  /// 正常完整句子
  case empty        = 2  /// 完整但内容为空
  case specialSymbol = 3 /// 完整但只有特殊符号
  case paragraphEnd = 4  /// 段落结束
}

// MARK: - SubtitleMsgData 验证扩展
extension SubtitleMsgData {

  /// 是否为机器人消息
  func isBotMessage(_ botUserId: String) -> Bool {
    return userId == botUserId
  }

  /// 是否为完整句子
  var isCompleteSentence: Bool {
    return definite
  }

  /// 是否为段落结束
  var isParagraphEnd: Bool {
    return paragraph
  }

  /// 文本是否为空或无效
  var isEmptyOrNil: Bool {
    return text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
  }

  /// 是否为特殊符号（~）
  var isSpecialSymbol: Bool {
    return text.trimmingCharacters(in: .whitespacesAndNewlines) == "~"
  }

  /// 是否需要特殊处理
  var needsSpecialHandling: Bool {
    return isSpecialSymbol
  }

  /// 判断机器人的完整句子是否需要扣费
  func isBotCompleteSentenceNeedsSpecialHandling(botUserId: String) -> Bool {
    guard isBotMessage(botUserId) else { return false }

    let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
    // 特殊字符不扣费
    if trimmed == "~" { return false }
    // 不是完整段落不扣费
    guard paragraph else { return false }
    return true
  }

  /// 判断机器人完整句子是否需要播放音频（特殊内容）
  func isBotCompleteSentenceAndPlayAudio(botUserId: String) -> Bool {
    guard isBotMessage(botUserId) else { return false }
    guard isCompleteSentence else { return false }
    return needsSpecialHandling
  }

  /// 获取句子类型
  var sentenceType: SubtitleSentenceType {
    if !isCompleteSentence { return .incomplete }
    if isParagraphEnd { return .paragraphEnd }
    if isEmptyOrNil { return .empty }
    if isSpecialSymbol { return .specialSymbol }
    return .complete
  }
}
